Mathlib/NumberTheory/ModularForms/JacobiTheta/Bounds.lean

```diff
@@ -0,0 +1,271 @@
+      + a * rexp (-π * a ^ 2 * t) / (1 - rexp (-π * t))) =O[atTop] fun t ↦ exp (-p * t) by
+    let ⟨p, hp, hp'⟩ := this
+    refine ⟨p, hp, (Eventually.isBigO ?_).trans hp'⟩
+    filter_upwards [eventually_gt_atTop 0] with t ht
+    exact F_nat_one_le ha ht
+  have aux : IsBigO atTop (fun t : ℝ ↦ (1 - rexp (-π * t))⁻¹) (fun _ ↦ (1 : ℝ)) := by
+    refine ((Tendsto.const_sub _ ?_).inv₀ (by norm_num)).isBigO_one ℝ (c := ((1 - 0)⁻¹ : ℝ))
+    simpa only [neg_mul, tendsto_exp_comp_nhds_zero, tendsto_neg_atBot_iff]
+      using tendsto_id.const_mul_atTop pi_pos
+  have aux' : IsBigO atTop (fun t : ℝ ↦ ((1 - rexp (-π * t)) ^ 2)⁻¹) (fun _ ↦ (1 : ℝ)) := by
+    simpa only [inv_pow, one_pow] using aux.pow 2
+  rcases eq_or_lt_of_le ha with rfl | ha'
+  · exact ⟨_, pi_pos, by simpa only [zero_pow two_ne_zero, zero_add, mul_one, zero_mul, zero_div,
+      add_zero] using (isBigO_refl _ _).mul aux'⟩
+  · refine ⟨π * a ^ 2, mul_pos pi_pos <| pow_pos ha' _, IsBigO.add ?_ ?_⟩
+    · conv_rhs => enter [t]; rw [← mul_one (rexp _)]
+      refine (Eventually.isBigO ?_).mul aux'
+      filter_upwards [eventually_gt_atTop 0] with t ht
+      rw [norm_of_nonneg (exp_pos _).le, exp_le_exp]
+      nlinarith [pi_pos]
+    · simp_rw [mul_div_assoc, ← neg_mul]
+      apply IsBigO.const_mul_left
+      simpa only [mul_one] using (isBigO_refl _ _).mul aux
+
+end k_eq_one
+
+end nat
+
+section int
+
+/-- Summand in the sum to be bounded (`ℤ` version). -/
+def f_int (k : ℕ) (a t : ℝ) (n : ℤ) : ℝ := |n + a| ^ k * exp (-π * (n + a) ^ 2 * t)
+
+lemma f_int_ofNat (k : ℕ) {a : ℝ} (ha : 0 ≤ a) (t : ℝ) (n : ℕ) :
+    f_int k a t (Int.ofNat n) = f_nat k a t n := by
+  rw [f_int, f_nat, Int.ofNat_eq_coe, Int.cast_natCast, abs_of_nonneg (by positivity)]
+
+lemma f_int_negSucc (k : ℕ) {a : ℝ} (ha : a ≤ 1) (t : ℝ) (n : ℕ) :
+    f_int k a t (Int.negSucc n) = f_nat k (1 - a) t n := by
+  have : (Int.negSucc n) + a = -(n + (1 - a)) := by { push_cast; ring }
+  rw [f_int, f_nat, this, abs_neg, neg_sq, abs_of_nonneg (by linarith)]
+
+lemma summable_f_int (k : ℕ) (a : ℝ) {t : ℝ} (ht : 0 < t) : Summable (f_int k a t) := by
+  apply Summable.of_norm
+  have := (HasSum.int_rec (summable_f_nat k a ht).hasSum
+    (summable_f_nat k (1 - a) ht).hasSum).summable.norm
+  convert this using 2 with n
+  rcases n with m | m
+  · simp only [f_int, f_nat, Int.ofNat_eq_coe, Int.cast_natCast, norm_mul, norm_eq_abs, abs_pow,
+      abs_abs]
+  · simp only [f_int, f_nat, Int.cast_negSucc, norm_mul, norm_eq_abs, abs_pow, abs_abs,
+      (by { push_cast; ring } : -↑(m + 1) + a = -(m + (1 - a))), abs_neg, neg_sq]
+
+lemma _root_.UnitAddCircle.eq_coe_Ico (a : UnitAddCircle) : ∃ b : ℝ, b ∈ Ico 0 1 ∧ ↑b = a := by
+  let b := (QuotientAddGroup.equivIcoMod zero_lt_one 0) a
+  exact ⟨b.1, by simpa only [zero_add] using b.2,
+    (QuotientAddGroup.equivIcoMod zero_lt_one 0).symm_apply_apply a⟩
+
+/-- The sum to be bounded (`ℤ` version). -/
+def F_int (k : ℕ) (a : UnitAddCircle) (t : ℝ) : ℝ :=
```

Comment:
I'd probably state the periodicity as a separate lemma and then use that in the definition.
But if inlining it does not lead to problems later, it is fine as is.

Reply:
This idiom seems to work quite well (and avoids namespace clutter) so I'd rather leave it as-is.